modules/serdes/include/hephaestus/serdes/protobuf/enums.h

```diff
@@ -0,0 +1,120 @@
+
+namespace internal {
+template <EnumType ProtoT>
+[[nodiscard]] auto getProtoPrefix() -> std::string {
+  const auto enum_type_name = magic_enum::enum_type_name<ProtoT>();
+
+  // Underscores indicate nested proto enums, no underscore means it's a top-level proto enum. Top level enums
+  // do not have a prefix.
+  if (const auto underscore_pos = std::find(enum_type_name.begin(), enum_type_name.end(), '_');
+      underscore_pos == enum_type_name.end()) {
+    return "";
+  }
+
+  // Nested enums have a prefix, and enum values are separated by an underscore:
+  // ClassName_EnumName_ENUM_VALUE.
+  return fmt::format("{}_", enum_type_name);
+}
+
+/// Convert between enums and their protobuf counterparts. The following convention is used:
+/// enum class FooExternalEnum : { BAR1, BAR2 };
+/// struct Foo {
+///   enum class InternalEnum : { BAR1, BAR2 };
+/// };
+/// will have a protobuf counterpart
+/// enum FooExternalEnum : { BAR1,  BAR2 };
+/// enum Foo_InternalEnum : { Foo_InternalEnum_BAR1, Foo_InternalEnum_BAR2 };
+template <EnumType T, EnumType ProtoT>
+[[nodiscard]] auto getAsProtoEnum(T e) -> ProtoT {
+  const auto proto_prefix = getProtoPrefix<ProtoT>();
+  auto proto_enum_name =
+      fmt::format("{}{}", proto_prefix, magic_enum::enum_name(e));  // ClassName_EnumName_ENUM_VALUE
+
+  auto proto_enum = magic_enum::enum_cast<ProtoT>(proto_enum_name);
+  heph::throwExceptionIf<heph::InvalidParameterException>(
+      !proto_enum.has_value(),
+      fmt::format("The proto enum does not contain the requested key {}. Proto enum keys are\n{}",
+                  proto_enum_name, utils::format::toString(magic_enum::enum_names<ProtoT>())));
+
+  return proto_enum.value();
+}
+
+template <EnumType T, EnumType ProtoT>
+[[nodiscard]] auto createEnumLookupTable() -> std::unordered_map<T, ProtoT> {
+  std::unordered_map<T, ProtoT> lookup_table;
+
+  for (const auto& e : magic_enum::enum_values<T>()) {
+    lookup_table[e] = getAsProtoEnum<T, ProtoT>(e);
+  }
+
+  return lookup_table;
+}
+
+/// @brief Create inverse lookup table. Unique values are guaranteed by the enum layout.
+template <EnumType T, EnumType ProtoT>
+[[nodiscard]] auto createInverseLookupTable(const std::unordered_map<T, ProtoT>& enum_to_proto_enum)
+    -> std::unordered_map<ProtoT, T> {
+  std::unordered_map<ProtoT, T> proto_enum_to_enum;
+
+  for (const auto& kvp : enum_to_proto_enum) {
+    proto_enum_to_enum.insert({ kvp.second, kvp.first });
+  }
+
+  return proto_enum_to_enum;
+}
+}  // namespace internal
+
+template <EnumType T, EnumType ProtoT>
+auto toProtoEnum(const T& enum_value) -> ProtoT {
+  static const auto enum_to_proto_enum = internal::createEnumLookupTable<T, ProtoT>();
+  try {
+    return enum_to_proto_enum.at(enum_value);
+  } catch (const std::out_of_range&) {
+    throw std::out_of_range(
+        fmt::format("Enum {} not found in the lookup table", magic_enum::enum_name(enum_value)));
+  }
```

Comment:
```suggestion
  throwExceptionIf<InvalidData>(!enum_to_proto_enum.contains(enum_value), "....");
    return enum_to_proto_enum[enum_value];
```

Reply:
Is that efficient though? This code is the one which is being called often, so it is performance critical. I was even thinking about removing the try catch block, as it is only there to extend the exception message by the enum value which is missing.

Wdyt? Remove try catch here alltogether?

Reply:
I mean if we really want to make it fast, we should do
```suggestion
  auto enum_it = enum_to_proto_enum.find(enum_value);
  throwExceptionIf<InvalidData>(enum_it == enum_to_proto_enum.end(), "....");
  return enum_it->second;
```